comtypes/_memberspec.py

```diff
@@ -173,62 +173,97 @@ def _fix_inout_args(
                 # this is the pointed-to type:
-                atyp = argtypes[i]._type_
+                atyp: Type[_CData] = getattr(argtypes[i], "_type_")
 
                 # Get the actual parameter, either as positional or
                 # keyword arg.
-                try:
-                    try:
-                        v = args[i]
-                    except IndexError:
-                        v = kw[name]
-                except KeyError:
-                    # no parameter was passed, make an empty one
-                    # of the required type
-                    v = atyp()
-                else:
-                    # parameter was passed, call .from_param() to
-                    # convert it to a ctypes type.
+
+                def prepare_parameter(v):
+                    # parameter was passed, call .from_param() to convert it to a ctypes type.
                     if getattr(v, "_type_", None) is atyp:
-                        # Array of or pointer to type 'atyp' was
-                        # passed, pointer to 'atyp' expected.
+                        # Array of or pointer to type 'atyp' was passed, pointer to 'atyp' expected.
                         pass
                     elif type(atyp) is SIMPLETYPE:
-                        # The from_param method of simple types
-                        # (c_int, c_double, ...) returns a byref()
-                        # object which we cannot use since later
-                        # it will be wrapped in a pointer.  Simply
-                        # call the constructor with the argument
-                        # in that case.
+                        # The from_param method of simple types (c_int, c_double, ...) returns a byref()
+                        # object which we cannot use since later it will be wrapped in a pointer.  Simply
+                        # call the constructor with the argument in that case.
                         v = atyp(v)
                     else:
                         v = atyp.from_param(v)
                         assert not isinstance(v, BYREFTYPE)
-                outargs[outnum] = v
-                outnum += 1
-                if len(args) > i:
-                    args[i] = v
-                else:
+                    return v
+
+                if is_positional:
+                    v = prepare_parameter(args[param_index])
+                    args[param_index] = v
+                elif name in kw:
+                    v = prepare_parameter(kw[name])
                     kw[name] = v
-            elif direction & 2 == 2:
+                else:
+                    # no parameter was passed, make an empty one of the required type
+                    # and pass it as a keyword argument
+                    v = atyp()
+                    if name is not None:
+                        kw[name] = v
+                    else:
+                        raise Exception("Unnamed inout parameters cannot be omitted")
```

Comment:
I'm not sure if it is more properly to raise `COMError` or maybe `TypeError` which seems used pretty often in similar cases.

After quick search across the whole repository I could find even such line:
```python
raise RuntimeError("BUG")
```
🤣

Reply:
I think it would be more appropriate to raise a `TypeError` here.

When throwing a `COMError`, I believe it should be raised as if it were thrown by the COM system of the execution environment (via CPython's `_ctypes`).
We can instantiate `_ctypes.COMError` by passing the error code (`hresult`), message, and a 5-length tuple as constructor arguments and then raise it, I am not sure about the correct combination of arguments in this case.
It seems to be a challenging decision within the community.

Instead, considering the situation as "unable to handle properly as a Python method type," I suggest raising a `TypeError`.

Reply:
> After quick search across the whole repository I could find even such line:
> 
> ```python
> raise RuntimeError("BUG")
> ```
> 
> 🤣

Oh, this is indeed one of those legacies that hasn't changed since this package was created.
In modern Python, carefully scrutinizing the types of errors thrown and crafting elegant messages can make the codebase more robust, effectively communicating the intentions of the implementer to future maintainers.